tests/src/test/java/org/eclipse/hono/tests/IntegrationTestSupport.java

```diff
@@ -999,29 +998,28 @@ public String setupGatewayDeviceBlocking(
             final Map<String, Object> properties,
             final long requestTimeout) {
 
-        return applicationClientFactory.getOrCreateCommandClient(tenantId).compose(commandClient -> {
-
-            commandClient.setRequestTimeout(requestTimeout);
-            final Promise<BufferResult> result = Promise.promise();
-            final Handler<Void> send = s -> {
-                // send the command upstream to the device
-                LOGGER.trace("sending command [name: {}, contentType: {}, payload: {}]", command, contentType, payload);
-                commandClient.sendCommand(deviceId, command, contentType, payload, properties).map(responsePayload -> {
-                    LOGGER.debug("successfully sent command [name: {}, payload: {}] and received response [payload: {}]",
-                            command, payload, responsePayload);
-                    return responsePayload;
-                }).recover(t -> {
-                    LOGGER.debug("could not send command or did not receive a response: {}", t.getMessage());
-                    return Future.failedFuture(t);
-                }).onComplete(result);
-            };
-            if (commandClient.getCredit() == 0) {
-                commandClient.sendQueueDrainHandler(send);
-            } else {
-                send.handle(null);
-            }
-            return result.future();
-        });
+        final Promise<Void> timeOutTracker = Promise.promise();
+        final long timerId = vertx.setTimer(requestTimeout, id -> timeOutTracker
+                .fail(new SendMessageTimeoutException("sending command timed out after " + requestTimeout + "ms")));
+
+        LOGGER.trace("sending command [name: {}, contentType: {}, payload: {}]", command, contentType, payload);
+        // send the command upstream to the device and receive the command response
+        final Future<BufferResult> sendCommandTracker = amqpApplicationClient
+                .sendCommand(tenantId, deviceId, command, contentType, payload, properties)
+                .onComplete(ar -> {
+                    vertx.cancelTimer(timerId);
+                    timeOutTracker.tryComplete();
+                    if (ar.succeeded()) {
+                        LOGGER.debug("successfully sent command [name: {}, payload: {}] and received response [payload: {}]",
```

Comment:
this is kind of misleading as it is also logged if the time out has already occurred, i.e. if the response arrives after the timeout, or am I mistaken?

Reply:
You are right. I have updated it now and also for other comments. Would you mind taking a look again?